### PR TITLE
[flutter_tools] generates version.json for web using flutter tool

### PR DESCRIPTION
**Pull request by balvinderz**
## Description

Generates version.json in web directory of project. version.json can be used in the future version of package_info which will support web. version.json has the following keys : app_name,version and build_number. This file is generated everytime a user runs or builds the project. 

## Related Issues

https://github.com/flutter/flutter/issues/46609 https://github.com/flutter/flutter/issues/63500
## Tests

I added the following tests:

Added a test that checks if the version.json was successfully generated or not.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for all changed/updated/fixed behaviors (See Contributor Guide).
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by balvinderz** [comment]
can you please review this . @jonahwilliams

**Comment by balvinderz** [comment]
can you please review this . @jonahwilliams

File: packages/flutter_tools/lib/src/commands/run.dart [hunk 1]
```diff
@@ -623,3 +627,31 @@ class RunCommand extends RunCommandBase {
     );
   }
 }
+Future<void> generateVersionFile(FlutterProject flutterProject) async {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/run.dart:630]: There is already code in the FlutterProject class to parse all of this information, why are you repeating the process yourself?

Comment (balvinderz) [reply packages/flutter_tools/lib/src/commands/run.dart:629]: didn't know . I m using it now in the new commit

**Comment by jonahwilliams** [comment]
This is getting close - I think currently the generation logic is not quite in the right place. For flutter run, I would say the actual version information is not that important, but if you wanted to generate it you should do it similar to how we generate sw.js:

https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/build_runner/devfs_web.dart#L784

For release builds, it should be done as part of a build target and not copied to web/, for example:

https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/build_system/targets/web.dart#L255

**Comment by balvinderz** [comment]
@jonahwilliams  i changed my approach . Instead of generating version.json i am injecting the version info in  index.html by creating a new script with id versionInfo . it has constants appName,buildNumber and version. This script tag is injected every time a user runs or builds the project.

File: packages/flutter_tools/lib/src/build_runner/devfs_web.dart [hunk 2]
```diff
@@ -7,6 +7,8 @@ import 'dart:typed_data';
 
 import 'package:dwds/data/build_result.dart';
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/build_runner/devfs_web.dart:8]: Please don't reformat the files

Comment (balvinderz) [reply packages/flutter_tools/lib/src/build_runner/devfs_web.dart:8]: i will revert

File: packages/flutter_tools/lib/src/build_runner/devfs_web.dart [hunk 3]
```diff
@@ -938,3 +942,30 @@ class ReleaseAssetServer {
     return shelf.Response.notFound('');
   }
 }
+void injectVersionScriptTag(
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/build_runner/devfs_web.dart:945]: I don't think this is the way to do it. If you want to insert it into the html, it should be as markup. Otherwise a JSON asset file seems like a reasonable way to go

Comment (balvinderz) [reply packages/flutter_tools/lib/src/build_runner/devfs_web.dart:945]: in one comment ditman suggested to inject it in a script tag .so i tried to do that.

Comment (balvinderz) [reply packages/flutter_tools/lib/src/build_runner/devfs_web.dart:945]: if that's not okay , i will revert to the previous way .

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_runner/devfs_web.dart:945]: if you inject it into a script tag you will need to use a javascript API to read it. Whereas you can use package:http JSON parse to read the version info in a json asset

**Comment by balvinderz** [comment]
@jonahwilliams  reverted to the old way.

File: packages/flutter_tools/lib/src/web/compile.dart [hunk 4]
```diff
@@ -111,28 +111,14 @@ class WebCompilationProxy {
   }
 }
 Future<void> generateVersionFile(FlutterProject flutterProject) async {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/web/compile.dart]: Could you avoid duplicating this logic twice?

File: packages/flutter_tools/test/general.shard/web/devfs_web_test.dart [hunk 5]
```diff
@@ -633,6 +634,7 @@ void main() {
     expect(webDevFS.webAssetServer.getFile('main.dart'), isNotNull);
     expect(webDevFS.webAssetServer.getFile('manifest.json'), isNotNull);
     expect(webDevFS.webAssetServer.getFile('flutter_service_worker.js'), isNotNull);
+    expect(webDevFS.webAssetServer.getFile('version.json'), isNotNull);
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/web/devfs_web_test.dart:637]: This isn't really ever testing that the contents are correct. Perhaps a unit test of your helper method to confirm it contains the correct details?

File: packages/flutter_tools/lib/src/build_system/targets/web.dart [hunk 6]
```diff
@@ -541,3 +545,14 @@ async function downloadOffline() {
...
+
+  final Map<String, String> versionFileJson = <String, String>{
+    'app_name': flutterProject.manifest.appName,
+    'version': flutterProject.manifest.buildName,
+    'build_number': flutterProject.manifest.buildNumber
+  };
+  environment.outputDir
+      .childFile('version.json')
+      .writeAsStringSync(jsonEncode(versionFileJson));
+}
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/build_system/targets/web.dart:558]: here and elsewhere: newline at EOF

File: packages/flutter_tools/lib/src/build_system/targets/web.dart [hunk 7]
```diff
@@ -266,6 +268,8 @@ class WebReleaseBundle extends Target {
         environment.outputDir.childFile(globals.fs.path.basename(outputFile.path)).path
       );
     }
+    generateVersionFile(FlutterProject.current(),environment);
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/build_system/targets/web.dart:271]: is this getting added to the output dependencies? Also, you need a test that this rule is doing the right thing

Comment (balvinderz) [reply packages/flutter_tools/lib/src/build_system/targets/web.dart:271]: yes. it is getting added to the output dependancies.

**Comment by balvinderz** [comment]
@jonahwilliams  will you review this please.

File: packages/flutter_tools/lib/src/base/utils.dart [hunk 8]
```diff
@@ -402,3 +403,12 @@ bool isWhitespace(_AnsiRun run) {
       rune == 0x3000 ||
       rune == 0xFEFF;
 }
+String getVersionInfo(FlutterProject flutterProject)  {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/base/utils.dart:406]: Rather than in utils, why not make this a method on `FlutterProject` itself?

File: packages/flutter_tools/lib/src/commands/run.dart [hunk 9]
```diff
@@ -622,4 +621,4 @@ class RunCommand extends RunCommandBase {
       endTimeOverride: appStartedTime,
     );
   }
-}
+}
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/run.dart:624]: here at elsewhere, add a newline at EOF

File: packages/flutter_tools/test/general.shard/build_system/targets/web_test.dart [hunk 10]
```diff
@@ -91,7 +91,18 @@ void main() {
     // Import.
     expect(generated, contains("import 'package:foo/main.dart' as entrypoint;"));
   }));
+  test('version.json is created after release build', () => testbed.run(() async {
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/build_system/targets/web_test.dart:94]: You can remove the `foo.txt` setup/assertion from this test, since it isn't what is being tested

Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/build_system/targets/web_test.dart:94]: add a newline between this test case and the previous one

File: packages/flutter_tools/test/general.shard/utils_test.dart [hunk 11]
```diff
@@ -21,7 +25,6 @@ baz=qux
       expect(file.values, hasLength(2));
     });
   });
-
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/utils_test.dart:24]: can you put this space back?

File: packages/flutter_tools/test/general.shard/utils_test.dart [hunk 12]
```diff
@@ -69,6 +72,21 @@ baz=qux
       expect(snakeCase('ABC'), equals('a_b_c'));
     });
   });
+  test('Version.json info is correct',(){
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/utils_test.dart:75]: Create the flutter manifest once, then pass it into the flutter project. You also should use a `MemoryFileSystem.test` instead of the globals.fs, like:


```
final MemoryFileSystem fileSystem = MemoryFileSystem.test();
final FlutterManifest manifest = FlutterManfiest.createFromString('''
... content goes here
''');
final FlutterProject flutterProject = FlutterProject(fileSystem.current, manifest, logger: BufferLogger.test());


```

File: packages/flutter_tools/test/general.shard/web/devfs_web_test.dart (hunk 5 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/general.shard/web/devfs_web_test.dart:637]: Thanks for adding that test, this is fine now

File: packages/flutter_tools/test/commands.shard/hermetic/build_web_test.dart [hunk 13]
```diff
@@ -161,7 +161,6 @@ class UrlLauncherPlugin {}
     ProcessManager: () => FakeProcessManager.any(),
     BuildSystem: () => MockBuildSystem(),
   });
-
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/commands.shard/hermetic/build_web_test.dart:164]: please add this space back

**Review by jonahwilliams — COMMENTED** [review]
Getting closer!

File: packages/flutter_tools/lib/src/build_runner/devfs_web.dart [hunk 14]
```diff
@@ -782,6 +782,7 @@ class WebDevFS implements DevFS {
       webAssetServer.writeBytes('stack_trace_mapper.js', stackTraceMapper.readAsBytesSync());
       webAssetServer.writeFile('manifest.json', '{"info":"manifest not generated in run mode."}');
       webAssetServer.writeFile('flutter_service_worker.js', '// Service worker not loaded in run mode.');
+      webAssetServer.writeFile('version.json',FlutterProject.current().getVersionInfo());
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/build_runner/devfs_web.dart:785]: spaces after `,`

File: packages/flutter_tools/lib/src/build_system/targets/web.dart [hunk 15]
```diff
@@ -266,6 +267,10 @@ class WebReleaseBundle extends Target {
         environment.outputDir.childFile(globals.fs.path.basename(outputFile.path)).path
       );
     }
+    final String versionInfo = FlutterProject.current().getVersionInfo();
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/build_system/targets/web.dart:270]: prefer 2 space indent

File: packages/flutter_tools/lib/src/commands/run.dart [hunk 16]
```diff
@@ -435,7 +435,6 @@ class RunCommand extends RunCommandBase {
     final bool hotMode = shouldUseHotMode();
 
     writePidFile(stringArg('pid-file'));
-
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/run.dart:438]: please revert this change

File: packages/flutter_tools/lib/src/project.dart [hunk 17]
```diff
@@ -263,6 +264,15 @@ class FlutterProject {
     }
     await injectPlugins(this, checkProjects: checkProjects);
   }
+  String getVersionInfo()  {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/project.dart:267]: space above and give this a doc comment describing what it does

File: packages/flutter_tools/lib/src/web/compile.dart [hunk 18]
```diff
@@ -105,4 +105,4 @@ class WebCompilationProxy {
   }) async {
     throw UnimplementedError();
   }
-}
+}
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/web/compile.dart:108]: newline at EOF

File: packages/flutter_tools/test/general.shard/build_system/targets/web_test.dart [hunk 19]
```diff
@@ -92,6 +92,16 @@ void main() {
...
   }));
 
+  test('version.json is created after release build', () => testbed.run(() async {
+    environment.defines[kBuildMode] = 'release';
+    final Directory webResources = environment.projectDir.childDirectory('web');
+    webResources.childFile('index.html')
+        .createSync(recursive: true);
+    environment.buildDir.childFile('main.dart.js').createSync();
+    await const WebReleaseBundle().build(environment);
+    expect(environment.outputDir.childFile('version.json'),exists);
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/build_system/targets/web_test.dart:102]:
```suggestion
    
    expect(environment.outputDir.childFile('version.json'),exists);
```

File: packages/flutter_tools/test/general.shard/project_test.dart [hunk 20]
```diff
@@ -241,6 +242,19 @@ void main() {
         expectExists(pluginRegistrantClasses.childFile('GeneratedPluginRegistrant.h'));
         expectExists(pluginRegistrantClasses.childFile('GeneratedPluginRegistrant.m'));
       });
+      testUsingContext('Version.json info is correct',(){
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/project_test.dart:245]:
```suggestion
      
      testUsingContext('Version.json info is correct',(){
```

File: packages/flutter_tools/test/general.shard/project_test.dart [hunk 21]
```diff
@@ -241,6 +242,19 @@ void main() {
...
+        final MemoryFileSystem fileSystem = MemoryFileSystem.test();
+        final FlutterManifest manifest = FlutterManifest.createFromString('''
+    name: test
+    version: 1.0.0+3
+    ''',logger: BufferLogger.test());
+        final FlutterProject project = FlutterProject(fileSystem.systemTempDirectory,manifest,manifest);
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/project_test.dart:251]: formatting please

**Review by jonahwilliams — COMMENTED** [review]
Please try and follow the general formatting of the rest of the code in the tool.

File: packages/flutter_tools/lib/src/project.dart [hunk 22]
```diff
@@ -263,6 +264,16 @@ class FlutterProject {
     }
     await injectPlugins(this, checkProjects: checkProjects);
   }
+
+  /// Returns a json encoded string that is used to generate version.json
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/project.dart:268]: `...containing the [appName], [version], and [buildNumber]`

File: packages/flutter_tools/test/general.shard/project_test.dart (hunk 20 above)
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/project_test.dart:245]:
```suggestion

      testUsingContext('Version.json info is correct', (){
```

File: packages/flutter_tools/test/general.shard/project_test.dart [hunk 23]
```diff
@@ -241,6 +242,19 @@ void main() {
         expectExists(pluginRegistrantClasses.childFile('GeneratedPluginRegistrant.h'));
         expectExists(pluginRegistrantClasses.childFile('GeneratedPluginRegistrant.m'));
       });
+      testUsingContext('Version.json info is correct',(){
+        final MemoryFileSystem fileSystem = MemoryFileSystem.test();
+        final FlutterManifest manifest = FlutterManifest.createFromString('''
+    name: test
+    version: 1.0.0+3
+    ''',logger: BufferLogger.test());
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/project_test.dart:250]:
```suggestion
        final FlutterManifest manifest = FlutterManifest.createFromString('''
    name: test
    version: 1.0.0+3
    ''', logger: BufferLogger.test());
```

File: packages/flutter_tools/test/general.shard/project_test.dart [hunk 24]
```diff
@@ -241,6 +242,19 @@ void main() {
...
+        final FlutterProject project = FlutterProject(fileSystem.systemTempDirectory,manifest,manifest);
+        final dynamic versionInfo = jsonDecode(project.getVersionInfo());
+        expect(versionInfo['app_name'],'test');
+        expect(versionInfo['version'],'1.0.0');
+        expect(versionInfo['build_number'],'3');
+
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/project_test.dart:256]:
```suggestion
```

File: packages/flutter_tools/test/general.shard/build_system/targets/web_test.dart [hunk 25]
```diff
@@ -92,6 +92,17 @@ void main() {
...
 
+  test('version.json is created after release build', () => testbed.run(() async {
+    environment.defines[kBuildMode] = 'release';
+    final Directory webResources = environment.projectDir.childDirectory('web');
+    webResources.childFile('index.html')
+        .createSync(recursive: true);
+    environment.buildDir.childFile('main.dart.js').createSync();
+    await const WebReleaseBundle().build(environment);
+
+    expect(environment.outputDir.childFile('version.json'),exists);
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/build_system/targets/web_test.dart:103]:
```suggestion
    expect(environment.outputDir.childFile('version.json'), exists);
```

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Comment by balvinderz** [comment]
@jonahwilliams testonly_devicelab_tests always fails on infra

**Comment by balvinderz** [comment]
thank you for merging and reviewing so many times @jonahwilliams

**Comment by jonahwilliams** [comment]
No problem!

**Comment by Leffe108** [comment]
Thank you both for your effort with this. 😊
